rust/src/rfb/rfb.rs

```diff
@@ -0,0 +1,851 @@
+                parser::RFBGlobalState::Message => {
+                    //todo implement RFB messages, for now we stop here
+                    return false;
+                }
+                _ => {
+                    SCLogDebug!("Invalid state for response");
+                    return false;
+                }
+            }
+        }
+    }
+
+    fn tx_iterator(
+        &mut self,
+        min_tx_id: u64,
+        state: &mut u64,
+    ) -> Option<(&RFBTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+}
+
+/// Probe to see if this input looks like a request or response.
+fn probe(input: &[u8]) -> bool {
+    // test for RFB
+    if input.len() > 4 && input[0] == 82 && input[1] == 70 && input[2] == 66 && input[3] == 32 {
+        return true;
+    }
+    return false;
+}
+
+// C exports.
+
+export_tx_get_detect_state!(
+    rs_rfb_tx_get_detect_state,
+    RFBTransaction
+);
+export_tx_set_detect_state!(
+    rs_rfb_tx_set_detect_state,
+    RFBTransaction
+);
+
+/// C entry point for a probing parser.
+#[no_mangle]
+pub extern "C" fn rs_rfb_probing_parser(
+    _flow: *const Flow,
+    _direction: u8,
+    input: *const u8,
+    input_len: u32,
+    _rdir: *mut u8
+) -> AppProto {
+    // Need at least 4 bytes.
+    if input_len > 4 && input != std::ptr::null_mut() {
+        let slice = build_slice!(input, input_len as usize);
+        if probe(slice) {
+            return unsafe { ALPROTO_RFB };
+        }
+    }
+    return ALPROTO_UNKNOWN;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_rfb_state_new() -> *mut std::os::raw::c_void {
+    let state = RFBState::new();
+    let boxed = Box::new(state);
+    return unsafe { transmute(boxed) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_rfb_state_free(state: *mut std::os::raw::c_void) {
+    // Just unbox...
+    let _drop: Box<RFBState> = unsafe { transmute(state) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_rfb_state_tx_free(
+    state: *mut std::os::raw::c_void,
+    tx_id: u64,
+) {
+    let state = cast_pointer!(state, RFBState);
+    state.free_tx(tx_id);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_rfb_parse_request(
+    _flow: *const Flow,
+    state: *mut std::os::raw::c_void,
+    pstate: *mut std::os::raw::c_void,
+    input: *const u8,
+    input_len: u32,
+    _data: *const std::os::raw::c_void,
+    _flags: u8,
+) -> i32 {
+    let eof = unsafe {
+        if AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF) > 0 {
+            true
+        } else {
+            false
+        }
+    };
+
+    if eof {
+        // If needed, handled EOF, or pass it into the parser.
+    }
```

Comment:
The template should probably change here..  This can probably changed to just return if input_len is 0. Checking for EOF is primarily useful if some cleanup needs to be done, like protocols supporting file extraction.

Reply:
would a return -1 be correct in this case?
If understanding logic in app-layer-parser.c correctly a return false would lead to disabling the applayer-parser for this flow....

Reply:
Returned 1 for now.